### ISSUE TITLE
Using an initramfs

### ISSUE THREAD

**Issue by fghsgh** [2bdkid/debian-rpi4-arm64-guide#2]
I figured out how to use an initramfs. You might consider adding this to the guide. This turned out to be really easy because the kernel, as compiled using this guide, already includes support for it, unlike the standard kernel that ships with Raspbian.

(either inside the chroot or on the system itself, I did it in the chroot because part of why I needed this was that the system was not booting, but that of course requires you not to delete `qemu-aarch64-static`)
1. `apt install initramfs-tools`
2. `update-initramfs -c` (if you're doing this from the chroot again, also add `-k <kernel version>`)
3. add `initramfs initrd.img-<kernel version> followkernel` to `config.txt`

Note that to get the kernel version when in a chroot, you can't do `uname -r`, because that gives you the kernel of the host, but instead you can do `ls /usr/lib/modules`.

You can also configure `update-initramfs` further as you would on a regular system, but for what I wanted (btrfs root filesystem) this turned out to be enough with the default configuration. (I did need to `apt install btrfs-progs`, and modify `cmdline.txt` to contain `rootfstype=btrfs`, but that's it and it depends on what you want the initramfs for)

**Comment by 2bdkid** [comment]
Awesome, thanks for looking into that. I'll give it a shot on my rpi.

**Comment by 2bdkid** [comment]
Did you try setting up the initrd from inside the chroot? Or only after the system was installed?

**Comment by fghsgh** [comment]
After the system was installed, from inside the chroot. I didn't delete `qemu-aarch64-static` because I knew I would want to enter the chroot in the future again. It should also work on the live system though.

**Comment by 2bdkid** [comment]
Ok, I'm working out steps to use dracut but it has trouble being installed in a chroot. It works fine on the running system tho. 

In your setup with the btrfs rootfs, is the initrd required because you need the btrfs module loaded early? Would it work without the initrd if btrfs was built into the kernel?

**Comment by fghsgh** [comment]
It was already built into the kernel actually, but for some reason it still didn't work (rebuilt it even with `fs/btrfs`, but it didn't help (it hanged on boot with no error messages and no shell), so my primary goal was to get an initramfs to have a shell to be able to inspect the situation, but that turned out to also fix the problem I was having).

To get the chroot to work better you might need to bind mount `/dev`, `/proc`, and `/sys`, that usually helps. Also run `su -` to initialize environment variables. I had a problem with `$PATH` being wrong and it failed at setting the locale (I should make an issue about that actually). Another problem was `sudo` complaining about the hostname.

**Comment by 2bdkid** [comment]
I'll give that a try. It was a dependency of dracut that wouldn't install correctly. It appears `bcm2711_defconfig` has btrfs built as a module, that may have been your problem.

**Comment by 2bdkid** [comment]
I pushed my dracut steps in 234ac887316e332aa9534b658172ff2f46e14d79. I know you already have your initrd working but if you find any issues with my steps just make another issue.
